py/server/deephaven/table_reader.py

```diff
@@ -0,0 +1,96 @@
+
+        2) The returned generator maintains a context internally and only guarantees the release of the context when
+            it is used in a for-loop statement. If the generator is used outside a for-loop statement by calling next()
+            on it, the context will not automatically be released. To avoid resource leaks, the generator should be
+            closed after use either by setting it to None or using the del statement or calling the close() method on
+            it.
+
+    Args:
+        table (Table): The table to read.
+        cols (Optional[Union[str, Sequence[str]]]): The columns to read. If None, all columns are read.
+        chunk_size (Optional[int]): The number of rows to read at a time. Default is None, meaning all rows in the table
+            are read in one chunk.
+
+    Returns:
+        A generator that yields a dictionary of column names to numpy arrays.
+    """
+    return _table_reader_by_row_set(table, cols, table.j_table.getRowSet(), chunk_size, False)
+
+
+def _table_reader_by_row_set(table: Table, cols: Optional[Union[str, Sequence[str]]], row_set: jpy.JType, chunk_size: Optional[int],
+                             prev: bool = False) -> Generator[Dict[str, np.ndarray], None, None]:
+    """ A generator that reads the chunks of rows from a table into a dictionary.  The dictionary is a map of column
+    names to numpy arrays.
+
+    Args:
+        table (Table):  The table to read.
+        cols (Optional[Union[str, Sequence[str]]]): The columns to read. If None, all columns are read.
+        row_set (jpy.JType): The row set to read.
+        chunk_size (Optional[int]): The number of rows to read at a time. If None, all rows in the row set are read.
+        prev (bool): If True, read the previous values. Default is False.
+
+    Returns:
+        A generator that yields a dictionary of column names to numpy arrays.
+    """
+    col_defs = _col_defs(table, cols)
+
+    row_sequence_iterator = row_set.getRowSequenceIterator()
+    col_sources = [table.j_table.getColumnSource(col_def.name) for col_def in col_defs]
+    chunk_size = row_set.size() if not chunk_size else chunk_size
+    j_reader_context = _JTableUpdateDataReader.makeContext(chunk_size, *col_sources)
+    with update_graph.auto_locking_ctx(table):
+        try:
+            while row_sequence_iterator.hasMore():
+                chunk_row_set = row_sequence_iterator.getNextRowSequenceWithLength(chunk_size)
+                j_array = _JTableUpdateDataReader.readChunkColumnMajor(j_reader_context, chunk_row_set, col_sources, prev)
+
+                col_dict = {}
+                for i, col_def in enumerate(col_defs):
+                    np_array = _column_to_numpy_array(col_def, j_array[i])
+                    col_dict[col_def.name] = np_array
+
+                yield col_dict
+        finally:
+            j_reader_context.close()
+            row_sequence_iterator.close()
```

Comment:
We are recommending that users do `gen=None`, `del gen`, or `gen.close()` if they are not use the gen in a for-loop.  At least the first relies on the GC getting around to deleting things.  Do any of these cases cause problems when garbage collection might not be immediate?  I don't remember the details of when it gets around to deleting things, but I seem to remember that it may be a long time before it does anything.

Reply:
Actually these recommended methods for clean up don't rely on Python generational GC to do the job (which is more for cyclic reference situations), but instead still on reference counting. So the dispose of the generator object is immediate and so is cleanup.